### PR TITLE
feat: physical names in the entire Construct Library

### PR DESCRIPTION
**Pull request by skinny85**
This PR changes the `xyzName` attribute of all Resources in the Construct Library to be of type PhysicalName,
and adds an AWS linter rule that checks ensures conformance.

The name of the property can be any ending substring of the base name of the class with the `Name`
suffix - for example, if my resource is `AwesomeFoobar`, the name can be `foobarName` or `awesomeFoobarName` (that's because we often have `Specialized1AwesomeFoobar` and `Specialized2AwesomeFoobar` in our library, that share a base prop interface).

There were a few interesting cases in the library which I didn't change, as I wasn't sure of the semantics of the resources, or they would require class name changes. Would appreciate some guidance there. These were:

* @aws-kms: AliasProps.name
* @aws-ssm: ParameterOptions.name
* @aws-applicationautoscaling: ScalableTargetProps.resourceId
* @aws-route53: CommonHostedZoneProps.zoneName
* @aws-route53: RecordSetOptions.recordName
* @aws-apigateway: StageProps.stageName
* @aws-servicediscovery: BaseNamespaceProps.name
* @aws-config: CustomRuleProps.ruleName
* @aws-config: ManagedRuleProps.ruleName
* @aws-config: AccessKeysRotatedProps.ruleName
* @aws-config: CloudFormationStackNotificationCheckProps.ruleName
* @aws-config: CloudFormationStackDriftDetectionCheckProps.ruleName
* @aws-rds: DatabaseClusterProps.databaseName / clusterIdentifier
* @aws-rds: DatabaseInstanceSourceProps.databaseName

BREAKING CHANGE: all resourceName attributes have their type changed from string to cdk.PhysicalName.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
Oops, I over-eagerly merged this because I thought it was ready and was trying to get ahead of merge conflicts.

Only later did I realize the PR was still under discussion, so I undid the merge. 

However, I can't seem to reopen the PR, so you're going to have to recreate it. 

Sorry!

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 1]
```diff
@@ -204,10 +204,12 @@ export class RestApi extends Resource implements IRestApi {
...
 
   constructor(scope: Construct, id: string, props: RestApiProps = { }) {
-    super(scope, id);
+    super(scope, id, {
+      physicalName: props.restApiName,
+    });
 
     const resource = new CfnRestApi(this, 'Resource', {
-      name: props.restApiName || id,
+      name: this.physicalName.value || id,
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: That looks a bit odd. The `|| id` means that the default behavior here is to allocate a physical name if nothing is assigned by the user, which is different than the normal behavior. I believe this is required by the RestAPI resource, and if I remember correctly, this is basically just a display name so doesn't have to be unique in any way or form.

I'd expect `this.physicalName.value` to return this value.

I wonder if the right approach should be to insert the default when calling to `super`:

```ts
super(scope, id, {
  physicalName: props.restApiName || PhysicalName.of(id)
}
```

File: packages/@aws-cdk/aws-apigateway/lib/vpc-link.ts [hunk 2]
```diff
@@ -37,14 +37,16 @@ export class VpcLink extends Resource {
...
 
   constructor(scope: Construct, id: string, props: VpcLinkProps) {
-    super(scope, id);
+    super(scope, id, {
+      physicalName: props.vpcLinkName,
+    });
 
     const cfnResource = new CfnVpcLink(this, 'Resource', {
-      name: props.name || this.node.uniqueId,
+      name: this.physicalName.value || this.node.uniqueId,
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/lib/vpc-link.ts]: same here. If the default behavior is to allocate a physical name, we need `this.physicalName` to return it.

File: packages/@aws-cdk/aws-cloudtrail/lib/index.ts [hunk 3]
```diff
@@ -178,7 +180,16 @@ export class Trail extends Resource {
...
+    const resourceIdentifiers = new ResourceIdentifiers(this, {
+      arn: trail.trailArn,
+      name: trail.trailName,
+      arnComponents: {
+        service: 'cloudtrail',
+        resource: 'trail',
+        resourceName: this.physicalName.value,
+      },
+    });
+    this.trailArn = resourceIdentifiers.arn;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudtrail/lib/index.ts]: How do we enforce using this and not just `trail.trailArn`? Seems voluntary... I am worried that most people will not know to do this.

File: packages/@aws-cdk/aws-codecommit/test/test.codecommit.ts [hunk 4]
```diff
@@ -42,8 +42,9 @@ export = {
     'fails when triggers have duplicate names'(test: Test) {
       const stack = new Stack();
 
-      const props = { repositoryName: 'MyRepository' };
-      const myRepository = new Repository(stack, 'MyRepository', props).notify('myTrigger');
+      const myRepository = new Repository(stack, 'MyRepository', {
+        repositoryName: PhysicalName.of('MyRepository'),
+      }).notify('myTrigger');
```
Comment (eladb) [inline packages/@aws-cdk/aws-codecommit/test/test.codecommit.ts]: `notify`? is this an API we like?

Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/test/test.codecommit.ts]: Maybe not, but I don't feel like it's related to the physical name change.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Yes. The problem here is that `this.id` is not available in the `super` call. We can change the value passed to `PhysicalName.of()` to be lazy, but that might actually change the semantic in a few cases (some tests, for example, assume that the `id` is a resolved value - not saying this particular case does it; I'll check).

File: packages/@aws-cdk/aws-cloudtrail/lib/index.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudtrail/lib/index.ts]: That's fine; adding this later is a backwards-compatible change.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Oh, sorry, I was thinking about a different case, where id = uniqueId. Will change.

Comment (skinny85) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Done.

File: packages/@aws-cdk/aws-apigateway/lib/vpc-link.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-apigateway/lib/vpc-link.ts]: Done.

**Comment by skinny85** [comment]
Moved to #2894 after Rico rekt the PR 🙂

**Pull request merged**